Data Race in Execute

Running tests with `go test -race -count 5 -parallel 32`, shows:

```
==================
WARNING: DATA RACE
Write at 0x00c000007138 by goroutine 104:
  github.com/graphql-go/graphql.Execute.func1()
      F:/Go/src/github.com/graphql-go/graphql/executor.go:61 +0x333

Previous read at 0x00c000007138 by goroutine 43:
  github.com/graphql-go/graphql.Execute()
      F:/Go/src/github.com/graphql-go/graphql/executor.go:75 +0x345
  github.com/graphql-go/graphql.Do()
      F:/Go/src/github.com/graphql-go/graphql/graphql.go:55 +0x664
  github.com/graphql-go/graphql_test.TestContextDeadline()
      F:/Go/src/github.com/graphql-go/graphql/executor_test.go:1895 +0x5c8
  testing.tRunner()
      C:/Go/src/testing/testing.go:865 +0x16a

Goroutine 104 (running) created at:
  github.com/graphql-go/graphql.Execute()
      F:/Go/src/github.com/graphql-go/graphql/executor.go:36 +0x215
  github.com/graphql-go/graphql.Do()
      F:/Go/src/github.com/graphql-go/graphql/graphql.go:55 +0x664
  github.com/graphql-go/graphql_test.TestContextDeadline()
      F:/Go/src/github.com/graphql-go/graphql/executor_test.go:1895 +0x5c8
  testing.tRunner()
      C:/Go/src/testing/testing.go:865 +0x16a
==================
--- FAIL: TestValidate_OverlappingFieldsCanBeMerged_ReturnTypesMustBeUnambiguous_ReportsCorrectlyWhenANonExclusiveFollosAnExclusive (0.02s)
```

--- a/executor.go
+++ b/executor.go
@@ -46,7 +46,7 @@ func Execute(p ExecuteParams) (result *Result) {
 		addExtensionResults(&p, result)
 	}()
 
-	resultChannel := make(chan *Result)
+	resultChannel := make(chan *Result, 2)
 
 	go func() {
 		result := &Result{}
@@ -55,10 +55,7 @@ func Execute(p ExecuteParams) (result *Result) {
 			if err := recover(); err != nil {
 				result.Errors = append(result.Errors, gqlerrors.FormatError(err.(error)))
 			}
-			select {
-			case resultChannel <- result:
-			case <-ctx.Done():
-			}
+			resultChannel <- result
 		}()
 
 		exeContext, err := buildExecutionContext(buildExecutionCtxParams{

--- a/executor.go
+++ b/executor.go
@@ -47,20 +47,20 @@ func Execute(p ExecuteParams) (result *Result) {
 	}()
 
 	resultChannel := make(chan *Result)
-	result = &Result{
-		Extensions: map[string]interface{}{},
-	}
 
-	go func(out chan<- *Result, done <-chan struct{}) {
+	go func() {
+		result := &Result{}
+
 		defer func() {
 			if err := recover(); err != nil {
-				result.AppendErrors(gqlerrors.FormatError(err.(error)))
+				result.Errors = append(result.Errors, gqlerrors.FormatError(err.(error)))
 			}
 			select {
-			case out <- result:
-			case <-done:
+			case resultChannel <- result:
+			case <-ctx.Done():
 			}
 		}()
+
 		exeContext, err := buildExecutionContext(buildExecutionCtxParams{
 			Schema:        p.Schema,
 			Root:          p.Root,
@@ -72,26 +72,26 @@ func Execute(p ExecuteParams) (result *Result) {
 		})
 
 		if err != nil {
-			result.AppendErrors(gqlerrors.FormatError(err))
+			result.Errors = append(result.Errors, gqlerrors.FormatError(err.(error)))
+			resultChannel <- result
 			return
 		}
 
-		result = executeOperation(executeOperationParams{
+		resultChannel <- executeOperation(executeOperationParams{
 			ExecutionContext: exeContext,
 			Root:             p.Root,
 			Operation:        exeContext.Operation,
 		})
-
-	}(resultChannel, ctx.Done())
+	}()
 
 	select {
 	case <-ctx.Done():
-		result.AppendErrors(gqlerrors.FormatError(ctx.Err()))
+		result := &Result{}
+		result.Errors = append(result.Errors, gqlerrors.FormatError(ctx.Err()))
+		return result
 	case r := <-resultChannel:
-		result = r
+		return r
 	}
-
-	return
 }
 
 type buildExecutionCtxParams struct {

--- a/extensions.go
+++ b/extensions.go
@@ -236,7 +236,7 @@ func addExtensionResults(p *ExecuteParams, result *Result) {
 			func() {
 				defer func() {
 					if r := recover(); r != nil {
-						result.AppendErrors(gqlerrors.FormatError(fmt.Errorf("%s.GetResult: %v", ext.Name(), r.(error))))
+						result.Errors = append(result.Errors, gqlerrors.FormatError(fmt.Errorf("%s.GetResult: %v", ext.Name(), r.(error))))
 					}
 				}()
 				if ext.HasResult() {

--- a/types.go
+++ b/types.go
@@ -1,8 +1,6 @@
 package graphql
 
 import (
-	"sync"
-
 	"github.com/graphql-go/graphql/gqlerrors"
 )
 
@@ -13,19 +11,9 @@ type Result struct {
 	Data       interface{}                `json:"data"`
 	Errors     []gqlerrors.FormattedError `json:"errors,omitempty"`
 	Extensions map[string]interface{}     `json:"extensions,omitempty"`
-
-	errorsLock sync.Mutex
 }
 
 // HasErrors just a simple function to help you decide if the result has errors or not
 func (r *Result) HasErrors() bool {
 	return len(r.Errors) > 0
-}
-
-// AppendErrors is the thread-safe way to append error(s) to Result.Errors.
-func (r *Result) AppendErrors(errs ...gqlerrors.FormattedError) {
-	r.errorsLock.Lock()
-	defer r.errorsLock.Unlock()
-
-	r.Errors = append(r.Errors, errs...)
 }

